File: CommandNavigation/CommandNavigationx2.cs
namespace CommandNavigation.Command1Navigation2 {
  using System;
  using System.Collections.Generic;
  /// <summary>
  /// Push:  -> Item -> Push in stack -> invoke OnPush -> invoke OnPushed event
  /// Pop:   -> invoke OnPop -> Pop form Stack -> invoke OnPopped event -> Item
  /// TryPop:-> Pop form Stack -> invoke OnPop -> invoke OnPopped event -> out Item
  /// </summary>
  /// <typeparam name="T"></typeparam>
  public class CommandNavigation<T> : Stack<T> where T : class, ICommandCtrl {
    public event OnCommandPoppedHandle<T> OnPopped;
    public event OnCommandPushedHandle<T> OnPushed;

    public CommandNavigation() : base() { }
    public CommandNavigation(int Capacity) : base(Capacity) { }

    public int CurrentOrder { get => Count == 0 ? int.MinValue : Peek().Order; }
    public int NextOrder { get => Count == 0 ? 0 : Peek().Order + 10; }

    public new void Clear() {
      while (Count > 0) {
        this.Pop();
      }
    }
    public void Discard() {
      base.Clear();
    }
    public new void Push(T Item) {
      if (Count == 0) {
        base.Push(Item);
        Item.OnPush();
        OnPushed?.Invoke(this, Item);
      }
      else if (Item.Order <= Peek().Order) {
        Pop();
        this.Push(Item);
      }
      else {
        base.Push(Item);
        Item.OnPush();
        OnPushed?.Invoke(this, Item);
      }
    }
    public new T Pop() {
      if (Count > 0) {
        var Popped = base.Pop();
        Popped.OnPop();
        OnPopped?.Invoke(this, Popped);
        return Popped;
      }
      throw new IndexOutOfRangeException("Empty Stack");
    }


  }

  public delegate void OnCommandPushedHandle<T>(CommandNavigation<T> CurrentStack, T PushedItem) where T : class, ICommandCtrl;
  public delegate void OnCommandPoppedHandle<T>(CommandNavigation<T> CurrentStack, T PoppedItem) where T : class, ICommandCtrl;

}

File: CommandNavigation.TS/src/CommandNavigationx2.ts
import { ICommandCtrl } from "./Interfaces";
import { Event2 } from "./../node_modules/katkits.ts/lib/Event"
export default class CommandNavigation extends Array<ICommandCtrl> {
  public readonly OnPopped: Event2<CommandNavigation, ICommandCtrl>;
  public readonly OnPushed: Event2<CommandNavigation, ICommandCtrl>;
  constructor() {
    super();
    this.OnPushed = new Event2();
    this.OnPopped = new Event2();
  }
  public CurrentOrder(): number { return this.length === 0 ? Number.MIN_SAFE_INTEGER : this.Peek().Order; }
  public NextOrder(): number { return this.length === 0 ? 0 : this.Peek().Order + 10; }

  public Peek(): ICommandCtrl { return this[this.length - 1]; }

  public Push(Item: ICommandCtrl): void {
    if (this.length === 0) {
      Item.OnPush();
      super.push(Item);
      this.OnPushed.Invoke(this, Item);
    }
    else if (Item.Order <= this.Peek().Order) {
      this.Pop();
      this.Push(Item);
    }
    else {
      Item.OnPush();
      super.push(Item);
      this.OnPushed.Invoke(this, Item);
    }
  }

  public Pop(): ICommandCtrl {
    if (this.length > 0) {
      this.Peek().OnPop();
      const Popped = super.pop();
      this.OnPopped.Invoke(this, Popped);
      return Popped;
    }
    throw Error("Empty Stack");
  }

  public Clear() {
    while (this.length > 0) {
      this.Pop();
    }
  }
  public Discard() { super.splice(0, this.length); }
}

